src/Access/KerberosInit.cpp

```diff
@@ -0,0 +1,213 @@
+};
+
+int KerberosInit::init(const String & keytab_file, const String & principal, const String & cache_name)
+{
+    auto log = &Poco::Logger::get("KerberosInit");
+    LOG_DEBUG(log,"Trying to authenticate to Kerberos v5");
+
+    krb5_error_code ret;
+
+    const char *deftype = nullptr;
+    int flags = 0;
+
+    if (!std::filesystem::exists(keytab_file))
+        throw Exception("Error keytab file does not exist", ErrorCodes::KERBEROS_ERROR);
+
+    memset(&k5, 0, sizeof(k5));
+    ret = krb5_init_context(&k5.ctx);
+    if (ret)
+        throw Exception("Error while initializing Kerberos 5 library", ErrorCodes::KERBEROS_ERROR);
+
+    if (!cache_name.empty())
+    {
+        ret = krb5_cc_resolve(k5.ctx, cache_name.c_str(), &k5.out_cc);
+        if (ret)
+            throw Exception("Error in resolving cache", ErrorCodes::KERBEROS_ERROR);
+        LOG_DEBUG(log,"Resolved cache");
+    }
+    else
+    {
+        // Resolve the default cache and get its type and default principal (if it is initialized).
+        ret = krb5_cc_default(k5.ctx, &defcache);
+        if (ret)
+            throw Exception("Error while getting default cache", ErrorCodes::KERBEROS_ERROR);
+        LOG_DEBUG(log,"Resolved default cache");
+        deftype = krb5_cc_get_type(k5.ctx, defcache);
+        if (krb5_cc_get_principal(k5.ctx, defcache, &defcache_princ) != 0)
+            defcache_princ = nullptr;
+    }
+
+    // Use the specified principal name.
+    ret = krb5_parse_name_flags(k5.ctx, principal.c_str(), flags, &k5.me);
+    if (ret)
+        throw Exception("Error when parsing principal name " + principal, ErrorCodes::KERBEROS_ERROR);
+
+    // Cache related commands
+    if (k5.out_cc == nullptr && krb5_cc_support_switch(k5.ctx, deftype))
+    {
+        // Use an existing cache for the client principal if we can.
+        ret = krb5_cc_cache_match(k5.ctx, k5.me, &k5.out_cc);
+        if (ret && ret != KRB5_CC_NOTFOUND)
+            throw Exception("Error while searching for cache for " + principal, ErrorCodes::KERBEROS_ERROR);
+        if (!ret)
+        {
+            LOG_DEBUG(log,"Using default cache: {}", krb5_cc_get_name(k5.ctx, k5.out_cc));
+            k5.switch_to_cache = 1;
+        }
+        else if (defcache_princ != nullptr)
+        {
+            // Create a new cache to avoid overwriting the initialized default cache.
+            ret = krb5_cc_new_unique(k5.ctx, deftype, nullptr, &k5.out_cc);
+            if (ret)
+                throw Exception("Error while generating new cache", ErrorCodes::KERBEROS_ERROR);
+            LOG_DEBUG(log,"Using default cache: {}", krb5_cc_get_name(k5.ctx, k5.out_cc));
+            k5.switch_to_cache = 1;
+        }
+    }
+
+    // Use the default cache if we haven't picked one yet.
+    if (k5.out_cc == nullptr)
+    {
+        k5.out_cc = defcache;
+        defcache = nullptr;
+        LOG_DEBUG(log,"Using default cache: {}", krb5_cc_get_name(k5.ctx, k5.out_cc));
+    }
+
```

Comment:
In our implementation we use krb5_cc_support_switch to verify if default cache of given type supports switching.  
First we try to find a cache for principal using krb5_cc_cache_match for example. In case of success at the end of our init code we make the cache primary by krb5_cc_switch command. 
In it case if we cannot find the cache for given principal then we use default cache.